lib/Dialect/AffineOps/AffineOps.cpp

```diff
@@ -1354,11 +1354,51 @@ struct AffineForLoopBoundFolder : public OpRewritePattern<AffineForOp> {
     return matchSuccess();
   }
 };
+
+// This is a pattern to canonicalize affine for op loop bounds.
+struct AffineForOpBoundCanonicalizer : public OpRewritePattern<AffineForOp> {
+  using OpRewritePattern<AffineForOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(AffineForOp forOp,
+                                     PatternRewriter &rewriter) const override {
+    SmallVector<Value *, 4> lbOperands(forOp.getLowerBoundOperands());
+    SmallVector<Value *, 4> ubOperands(forOp.getUpperBoundOperands());
+
+    auto lbMap = forOp.getLowerBoundMap();
+    auto ubMap = forOp.getUpperBoundMap();
+    auto prevLbMap = lbMap;
+    auto prevUbMap = ubMap;
+
+    canonicalizeMapAndOperands(&lbMap, &lbOperands);
+    canonicalizeMapAndOperands(&ubMap, &ubOperands);
+
+    // Any canonicalization change always leads to updated map(s).
+    bool canonicalized = false;
```

Comment:
Early exit if none has changed still looks preferable to me:
```
if (lbMap == prevLbMap && ubMap == prevUbMap)
  return matchFailure;
if (lbMap != prevLbMap)
  forOp.setLowerBound(lbOperands, lbMap);
if (ubMap != prevUbMap)
  forOp.setUpperBound(ubOperands, ubMap);
rewriter.updatedRootInPlace(forOp);
return matchSuccess();
```

Reply:
Done.